domain/miningmanager/mempool/mempool.go

```diff
@@ -527,9 +532,79 @@ func (mp *mempool) addTransaction(tx *consensusexternalapi.DomainTransaction, ma
 		return nil, err
 	}
 
+	err = mp.addTransactionToOrderedTransactionsByFeeRate(tx)
+	if err != nil {
+		return nil, err
+	}
+
 	return txDescriptor, nil
 }
 
+func (mp *mempool) findTxIndexInOrderedTransactionsByFeeRate(tx *consensusexternalapi.DomainTransaction) (int, error) {
+	if tx.Fee == 0 || tx.Mass == 0 {
+		return 0, errors.Errorf("findTxIndexInOrderedTransactionsByFeeRate expects a transaction with " +
+			"populated fee and as")
+	}
+	txID := consensushashing.TransactionID(tx)
+	txFeeRate := float64(tx.Fee) / float64(tx.Mass)
+
+	return sort.Search(len(mp.orderedTransactionsByFeeRate), func(i int) bool {
+		elementFeeRate := float64(mp.orderedTransactionsByFeeRate[i].Fee) / float64(mp.orderedTransactionsByFeeRate[i].Mass)
+		return elementFeeRate > txFeeRate ||
+			(elementFeeRate == txFeeRate &&
+				consensusexternalapi.LessOrEqual(
+					(*consensusexternalapi.DomainHash)(txID),
+					(*consensusexternalapi.DomainHash)(consensushashing.TransactionID(mp.orderedTransactionsByFeeRate[i])),
+				))
+	}), nil
+}
+
+func (mp *mempool) addTransactionToOrderedTransactionsByFeeRate(tx *consensusexternalapi.DomainTransaction) error {
+	index, err := mp.findTxIndexInOrderedTransactionsByFeeRate(tx)
+	if err != nil {
+		return err
+	}
+
+	if index == len(mp.orderedTransactionsByFeeRate) {
+		mp.orderedTransactionsByFeeRate = append(mp.orderedTransactionsByFeeRate, tx)
+	} else {
+		// copied and modified from https://stackoverflow.com/a/61822301/2413761
```

Comment:
Is there any particular reason as to why you've chosen to do the insert this way?
Please consider: https://github.com/golang/go/wiki/SliceTricks#insert